keps/sig-api-machinery/4988-serve-pagination-from-cache/README.md

```diff
@@ -0,0 +1,395 @@
+  leading to increased load on both the API server and etcd.
+- Ignoring the limit parameter defeats the purpose of pagination,
+  potentially returning massive responses that can overwhelm clients.
+
+This situation leads to unexpected behavior for users,
+where enabling pagination can degrade performance instead of improving it.
+
+### Goals
+
+- Improve the performance of paginated list requests to be comparable to non-paginated lists.
+- Make pagination behavior predictable and consistent.
+- Reduce load on etcd caused by paginated requests.
+
+### Non-Goals
+
+- Serve `resourceVersion="N"` request from watch cache
+- Support indices when paginating.
+- Eliminate all paginated list requests to etcd.
+
+## Proposal
+
+Leveraging the recent rewrite of the watchcache storage layer to use a B-tree
+(https://github.com/kubernetes/kubernetes/pull/126754), we propose to utilize
+B-tree snapshots to serve paginated lists.
+
+Mechanism:
+1. **Snapshot Creation:** When a paginated list request (with a limit parameter
+   set) is received, the API server will create a snapshot using the efficient
+   [Clone()](https://pkg.go.dev/github.com/google/btree#BTree.Clone) method.
+   This clone is a lazy copy, only duplicating necessary nodes, resulting in
+   minimal overhead.
+2. **Snapshot Storage:** The snapshot will be stored in memory, keyed by
+   resourceVersion.
+3. **Serving Subsequent Pages:** When a subsequent request with a continue token
+   arrives, the API server will:
+   - Extract the resourceVersion from the continue token.
+   - Retrieve the corresponding snapshot using the resourceVersion as the key.
+   - Use the snapshot to serve the requested page, or pass the request to etcd is snapshot is not available.
+4. **Snapshot Cleanup:** Snapshots will be subject to a Time-To-Live (TTL)
+   mechanism. We will reuse the existing watch event cleanup logic, which has a
+   75s TTL. This ensures that snapshots don't accumulate indefinitely.
+
+As still some pagination requests will be delegated to etcd, we will monitor the
+success rate by measuring the pagination cache hit vs miss ratio.
+
+Consideration: Should we start respecting the limit parameter?
+
+### Risks and Mitigations
+
+#### Client setting limit while not supporting pagination
+
+#### Memory overhead
+
+No, B-tree only store pointers the actual objects, not the object themselves.
+The objects are already cached to serve watch, so it should only add a small
+overhead for the B-tree structure itself, which is negligible compared to the
+size of the cached objects.
+
+#### Pagination request hitting another apiserver
+
+When connecting directly expectation is that a client will stay connected to the
+same apiserver. What happens when apiserver is behind loadbalancer?
+
+For setups with L4 loadbalancer apiserver can be configured with Goaway, which
+requests client reconnects periodically, however per request probability should
+be configured around 0.1%.
```

Comment:
Is there a reason for 0.1% here?

Reply:
0.1% is the recommended to configuration for GOAWAY https://github.com/kubernetes/kubernetes/pull/88567